clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2231,7 +2231,8 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
   case CK_UserDefinedConversion:
     return Visit(const_cast<Expr*>(E));
 
-  case CK_NoOp: {
+  case CK_NoOp:
+  case CK_HLSLArrayRValue: {
```

Comment:
It shouldn't be possible to hit this case; a value of ArrayParameterType isn't a scalar.